common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -408,14 +507,24 @@ public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) throws IOExc
       throw new IllegalArgumentException(
         String.format("The attempt id %s in this FinalizeShuffleMerge message does not match "
           + "with the current attempt id %s stored in shuffle service for application %s",
-          msg.appAttemptId, appShuffleInfo.attemptId, msg.appId));
+            msg.appAttemptId, appShuffleInfo.attemptId, msg.appId));
+    }
+    Map<Integer, Map<Integer, AppShufflePartitionInfo>> shuffleMergePartitions =
+        appShuffleInfo.partitions.get(msg.shuffleId);
+    Map<Integer, AppShufflePartitionInfo> shufflePartitions = shuffleMergePartitions.get(msg.shuffleMergeId);
+    if (shufflePartitions == STALE_SHUFFLE_PARTITIONS) {
+      throw new RuntimeException(String.format("Shuffle merge finalize request for shuffle %s"
+        + " with shuffleMergeId %s is %s", msg.shuffleId, msg.shuffleMergeId,
+          ErrorHandler.BlockPushErrorHandler.STALE_SHUFFLE_FINALIZE));
+    } else {
+      shuffleMergePartitions.put(msg.shuffleMergeId, FINALIZED_SHUFFLE_PARTITIONS);
+      appShuffleInfo.partitions.put(msg.shuffleId, shuffleMergePartitions);
     }
-    Map<Integer, AppShufflePartitionInfo> shufflePartitions =
-      appShuffleInfo.partitions.remove(msg.shuffleId);
```

Comment:
What is the rationale behind not removing old shuffle id's ? (if this is the latest shuffle).
How long are we planning to keep them around ? This is effectively a memory leak.

Having said that, if you do want to do this, rewrite code as:

```
AtomicReference<Map<Integer, AppShufflePartitionInfo>> shufflePartitions = new AtomicReference<>(null);
shuffleMergePartitions.compute(msg.shuffleMergeId, (id, map) -> {
  if (null == map) { ... }
  else if (FINALIZED_SHUFFLE_PARTITIONS == map) { ... }
  else {
    if (STALE_SHUFFLE_PARTITIONS == map) { ... } 
    shufflePartitions.set(map);
    return FINALIZED_SHUFFLE_PARTITIONS;
  }
} );
```

Reply:
No the idea is to remove older `shuffleMergeId`s, I missed to handle that.